packages/server-side-render/src/server-side-render.js

```diff
@@ -68,105 +79,73 @@ export class ServerSideRender extends Component {
 		// If httpMethod is 'POST', send the attributes in the request body instead of the URL.
 		// This allows sending a larger attributes object than in a GET request, where the attributes are in the URL.
 		const isPostRequest = 'POST' === httpMethod;
-		const urlAttributes = isPostRequest ? null : sanitizedAttributes;
+		const urlAttributes = isPostRequest
+			? null
+			: sanitizedAttributes ?? null;
 		const path = rendererPath( block, urlAttributes, urlQueryArgs );
-		const data = isPostRequest ? { attributes: sanitizedAttributes } : null;
+		const data = isPostRequest
+			? { attributes: sanitizedAttributes ?? null }
+			: null;
 
 		// Store the latest fetch request so that when we process it, we can
 		// check if it is the current request, to avoid race conditions on slow networks.
-		const fetchRequest = ( this.currentFetchRequest = apiFetch( {
+		const fetchRequest = ( fetchRequestRef.current = apiFetch( {
 			path,
 			data,
 			method: isPostRequest ? 'POST' : 'GET',
 		} )
-			.then( ( response ) => {
+			.then( ( fetchResponse ) => {
 				if (
-					this.isStillMounted &&
-					fetchRequest === this.currentFetchRequest &&
-					response
+					isMountedRef.current &&
+					fetchRequest === fetchRequestRef.current &&
+					fetchResponse
 				) {
-					this.setState( { response: response.rendered } );
+					setResponse( fetchResponse.rendered );
 				}
 			} )
 			.catch( ( error ) => {
 				if (
-					this.isStillMounted &&
-					fetchRequest === this.currentFetchRequest
+					isMountedRef.current &&
+					fetchRequest === fetchRequestRef.current
 				) {
-					this.setState( {
-						response: {
-							error: true,
-							errorMsg: error.message,
-						},
+					setResponse( {
+						error: true,
+						errorMsg: error.message,
 					} );
 				}
 			} ) );
+
 		return fetchRequest;
 	}
 
-	render() {
-		const response = this.state.response;
-		const {
-			className,
-			EmptyResponsePlaceholder,
-			ErrorResponsePlaceholder,
-			LoadingResponsePlaceholder,
-		} = this.props;
-
-		if ( response === '' ) {
-			return (
-				<EmptyResponsePlaceholder
-					response={ response }
-					{ ...this.props }
-				/>
-			);
-		} else if ( ! response ) {
-			return (
-				<LoadingResponsePlaceholder
-					response={ response }
-					{ ...this.props }
-				/>
-			);
-		} else if ( response.error ) {
-			return (
-				<ErrorResponsePlaceholder
-					response={ response }
-					{ ...this.props }
-				/>
-			);
+	const debouncedFetchData = useDebounce( fetchData, 500 );
+
+	// When the component unmounts, set isMountedRef to false. This will
+	// let the async fetch callbacks know when to stop.
+	useEffect(
+		() => () => {
+			isMountedRef.current = false;
+		},
+		[]
+	);
+
+	useEffect( () => {
+		// Don't debounce the first fetch. This ensures that the first render
+		// shows data as soon as possible
+		if ( prevProps === undefined ) {
+			fetchData();
+		} else if ( ! isEqual( prevProps, props ) ) {
+			debouncedFetchData();
 		}
+	} );
```

Comment:
Nitpicking: I think we can potentially tidy these up into just one effect. IMHO it's clearer and we can save the need for comparing `isEqual` on every render.

```js
const isMountedRef = useRef(false);

useEffect(() => {
  function fetchData() {
    // ...
  }

  // Initial request
  if (!isMountedRef.current) {
    fetchData();
  }

  isMountedRef.current = true;

  const debounceFetchData = debounce(fetchData, 500);
  
  // Debounced request
  debounceFetchData();

  return () => {
    isMountedRef.current = false;
  }
}, [...])
```

It's not tested so I might have missed something though.

Reply:
I don't think the suggested code would work:

- It would cause `fetchData` to be called twice on the first render.
- It would cause `debounceFetchData` to be called on every render, which would cause `response` to be set on every render, thus causing an infinite loop, if I had to guess. The comparison against the previous props exists to prevent this.

Reply:
Ahh, yeah, you're right, I missed that part. I still believe we can somehow skip the `isEqual` comparison though, but that's not very important anyway.

Reply:
There's a pretty good article I found recently discussing this kinda stuff:

https://www.benmvp.com/blog/object-array-dependencies-react-useEffect-hook/

In this case, the `useEffect` callback is dependent upon `attributes`, `block`, `urlQueryArgs`, and `httpMethod`; the first 3 of those are objects (meaning we can't just use shallow equality checks), and using destructuring on the first 2 won't work since their properties may themselves be objects.

As suggested in the article, a `useDeepCompareEffect` hook would be the best way to solve the problem, but that would probably only be a slight improvement over what I'm already doing, and since it introduces an external dependency (albeit a small one), I think it might be better suited to a separate PR.

Reply:
Yeah, I think this is good enough, and we're just refactoring it to hooks anyway. 👍